GrowLVForMicroshift: Change disk '/dev/vda' to '/dev/sda' for windows

praveenkumar opened the pull request:
In windows the hyperV identified disk as SCSI and disk partitions are `/dev/sda*` instead `/dev/vda*` and this makes the pvs/lvs commands to fail since the `/etc/lvm/devices/system.devices` file contain `/dev/vda*`.

```
$ sudo pvdisplay
  Devices file PVID BAUW2KTKhF3RE03h4JEMd1EImwC3cWd0 last seen on
  /dev/vda3 not found.
```

In this PR we are changing the `/dev/vda*` to `/dev/sda*` so that disk resize works as expected in windows also.


**Fixes:** Issue #3718

cfergeau reviewed (changes requested):
Shouldn't this be done in snc instead? 
I'd also add some hardening when getting `rootPart` from the VM, and in `runGrowPart`

praveenkumar commented:
> Shouldn't this be done in snc instead?

How we do it in snc, before creating the windows bundle?

cfergeau commented:
> > Shouldn't this be done in snc instead?
> 
> How we do it in snc, before creating the windows bundle?

yes, we can't do it before that. I'm also wondering if we could use entries from `/dev/disk/by-xxxx/` instead as their name won't depend on the virtualization stack.

gbraad reviewed (changes requested):
Looks OK from code.

This means the next call will get a result back and the grow will not fail...

However, can you make sure to err if that value is invalid, like empty? So what if this returns hda in another situation, instead of vda or sda? Continue execution with an error message?

praveenkumar commented:
> However, can you make sure to err if that value is invalid, like empty? So what if this returns hda in another situation, instead of vda or sda? Continue execution with an error message?

In case of openshift bundle we have following logic, may be same I can use here which will also capture in case value is empty so user not get unexpected panic.
 
```
rootPart = strings.TrimSpace(rootPart)
	if !strings.HasPrefix(rootPart, "/dev/vda") && !strings.HasPrefix(rootPart, "/dev/sda") {
		return "", fmt.Errorf("Unexpected root device: %s", rootPart)
	}
```

praveenkumar committed:
runGrowPart: Get the partition number using len of '/dev/.da'

With current solution `rootPart[len(rootPart)-1:]` if rootPart is
`/dev/vda12` then the partition come up as `1` instead of `12`. Current
patch make sure it provide correct partition info like `12`.

praveenkumar commented:
/hold

need to refactor.

praveenkumar committed:
growRootFileSystem: Use '/sysroot' option for xfs_growfs instead disk partition

The root partition for both microshift and openshift bundle is mounted
on `/sysroot` as xfs file system. So we can directly use this partition
to extend the size instead specifically provide the
device mapper path `/dev/rhel/root` (in case of microshift) or
block device partition `/dev/vda[n]` (in case of openshift).

This will help us to simplify same command for both types of bundles.

praveenkumar committed:
getrootPartition: Make it generic by using blkid

This PR changes, `getrootPartition` helper funtion to use `blkid` and
a bit of refactor to make use of this generic `getrootPartition`
helper function which now provides the right disk partition
for openshift/microshift bundle.

praveenkumar committed:
growLVForMicroshift: use root partition as devices option for pvs/lvs command

This PR make sure we are using correct device partition for all pvs/lvs and
lvextend command and this will help us fix issue with windows when lvm
system device file exist in the bundle with `/dev/vda3` as device name
and device id.

fixes: #3718

praveenkumar commented:
/unhold

gbraad commented on pkg/crc/machine/start.go:
```diff
@@ -196,7 +190,7 @@ func growLVForMicroshift(sshRunner *crcssh.Runner, lvFullName string) error {
 
 	// Get the size of root lv
-	sizeLV, _, err := sshRunner.RunPrivileged("Get the size of root logical volume", "/usr/sbin/lvs", "-S", fmt.Sprintf("lv_full_name=%s", lvFullName), "--noheadings", "--nosuffix", "--units", "b", "-o", "lv_size")
+	sizeLV, _, err := sshRunner.RunPrivileged("Get the size of root logical volume", "/usr/sbin/lvs", "-S", fmt.Sprintf("lv_full_name=%s", lvFullName), "--noheadings", "--nosuffix", "--units", "b", "-o", "lv_size", "--devices", rootPart)
```
"/sysroot" ?

gbraad commented on pkg/crc/machine/start.go:
```diff
@@ -171,22 +174,13 @@ func runGrowpart(sshRunner *crcssh.Runner, rootPart string) error {
 
 	// Get the size of volume group
-	sizeVG, _, err := sshRunner.RunPrivileged("Get the volume group size", "/usr/sbin/vgs", "--noheadings", "--nosuffix", "--units", "b", "-o", "vg_size")
+	sizeVG, _, err := sshRunner.RunPrivileged("Get the volume group size", "/usr/sbin/vgs", "--noheadings", "--nosuffix", "--units", "b", "-o", "vg_size", "--devices", rootPart)
```
"/sysroot" ?

Everywhere else in the PR this has been removed/changed. I am not sure why that is... as it seems still relevant.

gbraad commented on pkg/crc/machine/start.go:
```diff
@@ -212,7 +206,7 @@ func growLVForMicroshift(sshRunner *crcssh.Runner, lvFullName string) error {
 	if sizeToIncrease > 1 {
 		logging.Info("Extending and resizing '/dev/rhel/root' logical volume")
-		if _, _, err := sshRunner.RunPrivileged("Extending and resizing the logical volume(LV)", "/usr/sbin/lvextend", "-L", fmt.Sprintf("+%db", sizeToIncrease), lvPath); err != nil {
+		if _, _, err := sshRunner.RunPrivileged("Extending and resizing the logical volume(LV)", "/usr/sbin/lvextend", "-L", fmt.Sprintf("+%db", sizeToIncrease), lvPath, "--devices", rootPart); err != nil {
```
"/sysroot" ?

gbraad reviewed (changes requested):
I see `rootPart` and `"/sysroot"` being used. Are they the same? And if so, please use a constant.

gbraad commented on pkg/crc/machine/start.go:
```diff
@@ -137,19 +132,27 @@ func growRootFileSystem(sshRunner *crcssh.Runner, preset crcPreset.Preset) error
 		return err
 	}
-	if _, _, err = sshRunner.RunPrivileged(fmt.Sprintf("Growing %s filesystem", rootPart), "xfs_growfs", rootPart); err != nil {
+	if _, _, err = sshRunner.RunPrivileged(fmt.Sprintf("Growing %s filesystem", rootPart), "xfs_growfs", "/sysroot"); err != nil {
```
Previous comments are regarding "`/sysroot` which seems to be a shared mount. seeing the change looks awkward without ulling this out as a more meaningful constant, like `rootMount`

praveenkumar replied:
Updated with `rootFS` variable.

gbraad commented on pkg/crc/machine/start.go:
```diff
@@ -137,19 +132,27 @@ func growRootFileSystem(sshRunner *crcssh.Runner, preset crcPreset.Preset) error
 	if _, _, err := sshRunner.RunPrivileged("Remounting /sysroot read/write", "mount -o remount,rw /sysroot"); err != nil {
```
you also use `rootMount` here, which justifies re-use for clarity

praveenkumar replied:
updated with `rootFS` variable
